server/src/main/java/org/elasticsearch/rest/action/ingest/RestSimulateIngestAction.java

```diff
@@ -0,0 +1,180 @@
+
+import static org.elasticsearch.rest.RestRequest.Method.GET;
+import static org.elasticsearch.rest.RestRequest.Method.POST;
+
+/**
+ * This is the REST endpoint for the simulate ingest API. This API executes all pipelines for a document (or documents) that would be
+ * executed if that document were sent to the given index. The JSON that would be indexed is returned to the user, along with the list of
+ * pipelines that were executed. The API allows the user to optionally send in substitute definitions for pipelines so that changes can be
+ * tried out without actually modifying the cluster state.
+ */
+@ServerlessScope(Scope.PUBLIC)
+public class RestSimulateIngestAction extends BaseRestHandler {
+
+    @Override
+    public List<Route> routes() {
+        return List.of(new Route(GET, "/_ingest/_simulate"), new Route(POST, "/_ingest/_simulate"));
+    }
+
+    @Override
+    public String getName() {
+        return "ingest_simulate_ingest_action";
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public RestChannelConsumer prepareRequest(final RestRequest request, final NodeClient client) throws IOException {
+        if (request.getRestApiVersion() == RestApiVersion.V_7 && request.hasParam("type")) {
+            request.param("type");
+        }
+        SimulateBulkRequest bulkRequest = new SimulateBulkRequest();
+        String defaultIndex = request.param("index");
+        FetchSourceContext defaultFetchSourceContext = FetchSourceContext.parseFromRestRequest(request);
+        String defaultPipeline = request.param("pipeline");
+        Tuple<XContentType, BytesReference> sourceTuple = request.contentOrSourceParam();
+        Map<String, Object> sourceMap = XContentHelper.convertToMap(sourceTuple.v2(), false, sourceTuple.v1()).v2();
+        bulkRequest.setPipelineSubstitutions((Map<String, Map<String, Object>>) sourceMap.remove("pipeline_substitutions"));
+        BytesReference transformedData = convertToBulkRequestXContentBytes(sourceMap);
+        bulkRequest.add(
+            transformedData,
+            defaultIndex,
+            null,
+            defaultFetchSourceContext,
+            defaultPipeline,
+            null,
+            true,
+            true,
+            request.getXContentType(),
+            request.getRestApiVersion()
+        );
+        return channel -> client.execute(SimulateBulkAction.INSTANCE, bulkRequest, new SimulateIngestRestToXContentListener(channel));
+    }
+
+    /*
+     * The simulate ingest API is intended to have inputs and outputs that are formatted similarly to the simulate pipeline API for the
+     * sake of consistency. But internally it uses the same code as the _bulk API, so that we have confidence that we are simulating what
+     * really happens on ingest. This method transforms simulate-style inputs into an input that the bulk API can accept.
+     */
+    private BytesReference convertToBulkRequestXContentBytes(Map<String, Object> sourceMap) throws IOException {
```

Comment:
At a glance (I haven't double checked, sorry), this looks like it could be `static` and then unit tested? Would it be worth doing that?

Reply:
Yep, I had meant to do that. I just pushed a couple of unit tests for this class.